bench/src/ffi/hs.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use std::cell::RefCell;
+use std::vec::IntoIter;
+
+use hyperscan::Error;
+use hyperscan::{BlockDatabase, BlockScanner, DatabaseBuilder, HS_CPU_FEATURES_AVX2, Pattern,
+                RawScratch, ScratchAllocator, HS_FLAG_ALLOWEMPTY, HS_FLAG_SINGLEMATCH,
+                HS_FLAG_SOM_LEFTMOST};
+
+/// Regex wraps a hyperscan regular expression.
+///
+/// It can be used safely from multiple threads simultaneously.
+pub struct Regex {
+    db: BlockDatabase,
+    s: RefCell<RawScratch>,
+}
+
+unsafe impl Send for Regex {}
```

Comment:
Is this actually necessary?

Reply:
Yes, seems some benchmark depend on the `Send` trait
>error[E0277]: the trait bound `*mut hyperscan::raw::hs_scratch: std::marker::Send` is not satisfied in `hyperscan::RawScratch`
  --> bench/src/misc.rs:32:1
   |
32 |    bench_match!(literal, r"y", {
   |   _^
   |  |_|
   | ||
33 | ||    format!("{}y", repeat("x").take(50).collect::<String>())
34 | || });
   | ||   ^
   | ||___|
   | |____`*mut hyperscan::raw::hs_scratch` cannot be sent between threads safely
   |      in this macro invocation
   |
   = help: within `hyperscan::RawScratch`, the trait `std::marker::Send` is not implemented for `*mut hyperscan::raw::hs_scratch`
   = note: required because it appears within the type `hyperscan::RawScratch`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::cell::RefCell<hyperscan::RawScratch>`
   = note: required because it appears within the type `ffi::hs::Regex`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::sync::Mutex<ffi::hs::Regex>`
   = note: required by `lazy_static::lazy::Lazy`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)